src/core/server/saved_objects/migrations/core/document_migrator.ts

```diff
@@ -67,21 +67,17 @@ import Semver from 'semver';
 import { Logger } from '../../../logging';
 import { RawSavedObjectDoc } from '../../serialization';
 import { SavedObjectsMigrationVersion } from '../../types';
-import { MigrationLogger, SavedObjectsMigrationLogger } from './migration_logger';
+import { MigrationLogger } from './migration_logger';
+import { SavedObjectTypeRegistry } from '../../saved_objects_type_registry';
+import { SavedObjectMigrationFn } from '../types';
 
 export type TransformFn = (doc: RawSavedObjectDoc) => RawSavedObjectDoc;
 
-type MigrationFn = (doc: RawSavedObjectDoc, log: SavedObjectsMigrationLogger) => RawSavedObjectDoc;
-
 type ValidateDoc = (doc: RawSavedObjectDoc) => void;
 
-export interface MigrationDefinition {
-  [type: string]: { [version: string]: MigrationFn };
-}
-
-interface Opts {
+interface DocumentMigratorOptions {
   kibanaVersion: string;
-  migrations: MigrationDefinition;
+  typeRegistry: SavedObjectTypeRegistry;
```

Comment:
Do you anticipate any issues with DocumentMigrator requiring the whole registry rather than just the migrations by type? I just worry about tight-coupling in an already complex area of Core. That said, maybe things are made simpler by having a single "config" type that represents all of the information about the registered types.

I ask because I'm not intimately familiar with the SO service and how easy/hard making changes in this area are. But if this PR is any indication, it appears that major changes require changing **many** files, which could be a code smell.

Reply:
> that said, maybe things are made simpler by having a single "config" type that represents all of the information about the registered types.

This is the idea. The `typeRegistry` should be the source of truth, and only responsible for all SO-registered things.

However, I may have been a little 'hard' on forcing everything to use the `typeRegistry`. Maybe adding methods on it to retrieve the specialized types for some or every usages is better.

Could add a `SavedObjectTypeRegistry.getMigrationDefinitions()` that returns the 'old' `MigrationDefinition`. Would that makes sense to you?

Reply:
I think it's a judgement call. If you think that would simplify testing or changing some SO components in the future, then 👍, if it's negligible then I think it's fine to keep as is. It may be that the internal organization doesn't fall along these lines well now, in which case it may not make sense to change anything.

Reply:
imho keeping the `typeRegistry` as the SSoT and passing it though the whole chain is probably more maintainable mid/long term. The code adaptions being already done, I think I would keep this unchanged if it's alright.